app/main.py

```diff
@@ -2,5 +2,20 @@
 
 
 def cache(func: Callable) -> Callable:
-    # Write your code here
-    pass
+    cache_storage = {}
+
+    def wrapper(*args, **kwargs) -> Callable:
+        if any(isinstance(arg, (list, dict, set)) for arg in args):
+            raise ValueError("All arguments must be immutable")
+        key = (func, args, frozenset(kwargs.items()))
```

Comment:
Using `frozenset` to convert `kwargs.items()` into an immutable type is correct, but ensure that all values in `kwargs` are hashable. If any value is a mutable type, it will raise an error.